collie/model/hybrid_matrix_factorization.py

```diff
@@ -207,38 +239,102 @@ def __init__(self,
+                        'parameter_prefix_list': [
+                            'item_metadata', 'combined', 'user_bias', 'item_bias'
+                        ],
+                        'stage': 'item_metadata_only',
+                    },
+                    {
+                        'lr': metadata_only_stage_lr,
+                        'optimizer': metadata_only_stage_optimizer,
+                        # optimize metadata layers only
+                        'parameter_prefix_list': [
+                            'user_metadata', 'combined', 'user_bias', 'item_bias'
+                        ],
+                        'stage': 'user_metadata_only',
+                    },
+                    {
+                        'lr': all_stage_lr,
+                        'optimizer': all_stage_optimizer,
+                        # optimize everything
+                        'parameter_prefix_list': [
+                            'user', 'item', 'item_metadata', 'user_metadata', 'combined'
+                        ],
+                        'stage': 'all',
+                    },
+                ]
+            elif user_metadata is not None:
+                optimizer_config_list = initial_optimizer_block + [
+                    {
+                        'lr': metadata_only_stage_lr,
+                        'optimizer': metadata_only_stage_optimizer,
+                        # optimize metadata layers only
+                        'parameter_prefix_list': [
+                            'user_metadata', 'combined', 'user_bias', 'item_bias'
+                        ],
+                        'stage': 'user_metadata_only',
+                    },
+                    {
+                        'lr': all_stage_lr,
+                        'optimizer': all_stage_optimizer,
+                        # optimize everything
+                        'parameter_prefix_list': ['user', 'item', 'user_metadata', 'combined'],
+                        'stage': 'all',
+                    },
+                ]
+            elif item_metadata is not None:
+                optimizer_config_list = initial_optimizer_block + [
+                    {
+                        'lr': metadata_only_stage_lr,
+                        'optimizer': metadata_only_stage_optimizer,
+                        # optimize metadata layers only
+                        'parameter_prefix_list': [
+                            'item_metadata', 'combined', 'user_bias', 'item_bias'
+                        ],
+                        'stage': 'item_metadata_only',
+                    },
+                    {
+                        'lr': all_stage_lr,
+                        'optimizer': all_stage_optimizer,
+                        # optimize everything
+                        'parameter_prefix_list': ['user', 'item', 'item_metadata', 'combined'],
+                        'stage': 'all',
+                    },
+                ]
```

Comment:
[Optional, super nitpicky] Can we add a line here to make it a bit visually clearer that the `if/elif` block has ended 😅

Reply:
👍